### PR TITLE
feat(seo): add per-route metadata, sitemap, robots, JSON-LD

### LINKED ISSUE #503: [Perf] Improve SEO
## Overview

An SEO audit of `afterclass-io/afterclass.io` revealed critical crawlability and indexation gaps across the Next.js App Router setup. Because dynamic course and professor pages are built entirely using parallel route slots (`@reviews`, `@header`, etc.) without route-segment `page.tsx` or `layout.tsx` files, thousands of UGC pages inherit the exact same generic root metadata (`title: "AfterClass"`). Additionally, the site lacks standard discoverability files (`sitemap.ts`, `robots.ts`), canonical URL tags, and structured data (JSON-LD).

---

### 🔴 Critical
* **C1 — Missing `sitemap.ts` & `robots.ts`:** No sitemap or robots file exists, causing search engines to miss dynamically generated course and professor routes while crawling private or API endpoints.
* **C2 — Duplicate Titles & Metadata Site-Wide:** Slot files cannot export route segment metadata. As a result, ~12k+ UGC reviews across thousands of course/professor URLs share the single static title `"AfterClass"` and generic site description.
* **C3 — Missing Canonical URLs (`alternates.canonical`):** Filter and search query parameters (`?course=...&filter=...&sort=...`) create duplicate content candidates without canonical tags.
* **C4 — Preview URL Leak in `metadataBase`:** `src/app/layout.tsx` prioritizes `process.env.VERCEL_URL` over production domain configs, causing preview domains to leak into canonical/OG URLs.

### 🟠 High Priority
* **H1 — Viewport Blocks Zoom:** `maximumScale: 1.0` in root viewport violates WCAG 1.4.4 and hurts Lighthouse accessibility/SEO scores.
* **H2 — Missing Structured Data (JSON-LD):** No Schema.org markup (`Course`, `Person`, `AggregateRating`, `Review`, `BreadcrumbList`, `WebSite`) despite having the underlying review and rating metadata available.
* **H3 — Incomplete Open Graph & Missing Twitter Cards:** Root OG lacks image arrays/URLs and Twitter card metadata; default locale is set to `en_US` rather than `en_SG`.
* **H4 — Client-Only Breadcrumbs:** Breadcrumb components are `"use client"` and render empty in server HTML without semantic structured data.
* **H5 — Indexable Private / Search Routes:** `/timetable`, `/roadmaps/mine`, `/submit`, `/account/auth/*`, and `/search?q=...` are indexable by default.
* **H6 — Missing Title Template:** No `title.template` in root metadata, forcing full title duplication on sub-pages.
* **H7 — Generic Image Alt Text:** Fallback alt tags (e.g., `alt="Announcement"`, `"AfterClass"`).

### 🟡 Medium & Low Priority
* **M2 / M8 — Crawlability & Internal Linking:** Lack of dedicated `/courses` or `/professors` index/directory pages; course anchors use only codes without descriptive course names.
* **M3 — Lack of ISR / Static Optimization:** All pages are fully dynamic; top courses/professors could leverage `generateStaticParams` / ISR revalidation.
* **M4 / M5 — Headers & Redirects:** `/reviews -> /` is temporary (307) rather than permanent (308); missing `X-Robots-Tag: noindex` for preview deployments and basic security/trust headers.
* **M9 — Missing Review Permalinks:** Reviews rely on query parameters that push unauthenticated users to login instead of crawlable public permalinks.
* **L1–L4 — Minor Polish:** `not-found.tsx` missing title metadata, Poppins font loads unused weights, missing sitelinks searchbox JSON-LD.

---

## Existing Strengths (Do Not Regress)
* File-based root OG/favicon conventions (`favicon.ico`, `opengraph-image.png`).
* Next.js Google font optimizations (`next/font/google` self-hosted with `subsets`).
* Consistent single `h1` pattern with `PageTitle` component.
* Server caching pattern established in `(school)/roadmaps/[id]/page.tsx` (`generateMetadata` + React `cache()`).

### LINKED ISSUE #524: [SEO] P0: Verify the slot-metadata premise on a live run
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Further Notes — the runtime assumption) · Source audit: `.scratch/reports/seo.md`

## What to build

The per-route metadata plan rests on one runtime claim that could not be confirmed against a live response: that a `generateMetadata` exported from a named parallel-route slot reaches the rendered head. It was verified against the installed framework's own metadata resolver source, but production sits behind a bot challenge that returns a challenge page to scripted requests.

Run the site locally and read the head of a course page from the server's HTML. If the link-preview tag resolves to `/course/<code>/opengraph-image-<hash>`, the premise holds and the metadata plan is additive as written. If it does not, the premise is wrong and the per-route metadata ticket is architectural rather than additive — record the outcome and stop that sibling for re-planning rather than forcing it through.

Everything else in the specification is independent of this answer and proceeds regardless.

## Acceptance criteria

- [ ] Local run; the head of a course page read from the raw server HTML (not the hydrated DOM)
- [ ] Outcome recorded on this ticket, with the observed tag
- [ ] If the premise fails: the finding posted on the per-route metadata sibling under #503 so it is re-planned, not forced

## Blocked by

- None (can start immediately)

### LINKED ISSUE #525: [SEO] P0: Point the metadata base URL at the validated site URL
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — The metadata base URL)

## What to build

Every canonical and preview URL the site emits names the wrong host — in production, not just on previews. The root layout prefers the raw per-deployment deployment hostname over the validated site URL the environment module already resolves correctly, and that variable is set on production deployments too, so production pages point at a per-deployment host rather than the site domain.

The layout reads the validated value and the raw access is deleted. This is one line, and it is first in the tier because every other absolute URL in it depends on the base being right.

Verification is by inspection plus one check on a preview deployment: CI runs against localhost, where no deployment hostname is set, so the suite can assert the *path* of a canonical or preview URL but never the *host* — which is the whole content of this finding.

## Acceptance criteria

- [ ] Root metadata base read from the validated environment value; the raw deployment-hostname access deleted
- [ ] No raw read of that variable remains anywhere a validated value exists
- [ ] Verified on a preview deployment — production URLs name the site domain; a localhost run is not sign-off for a host-resolution change
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- None (can start immediately)

### LINKED ISSUE #526: [SEO] P0: Give course, professor and home pages their own head
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Per-route metadata; The parallel-route correction)

## What to build

A student searching "IS216 review" gets a result titled with the course and described by that course's own reviews, instead of the site name and the same three sentences served on every one of thousands of course and professor URLs.

Course, professor and home pages each gain a `generateMetadata` in the `@reviews` slot that already owns the link-preview image: a title naming the course or professor, and a description composed from that page's own figures — average rating, review count, label breakdown — through the existing public procedures. Descriptions are computed from that data rather than written as static strings, which is what makes them distinct per URL.

The fetch is wrapped in the request-scoped cache the roadmap detail route already proves, so the metadata call and the render share one query per request. The root title becomes a template so per-route titles append the brand once, in one place. A missing course or professor returns a not-found title rather than throwing.

**The one constraint a reviewer must check: exactly one slot may own metadata for a route — `@reviews`.** Metadata items merge in traversal order and the last writer wins, so two slots both exporting metadata makes the winner a function of the order the framework happens to walk the tree. `@header` is the tempting alternative — it already fetches the records — but splitting the image and the tags across two slots is exactly the arrangement this constraint exists to prevent. The constraint is recorded as a comment at each owning slot, because it is invisible in the file that would violate it.

This ticket also creates the shared end-to-end seam every later ticket in this set adds rows to: a table-driven spec asserting against the raw HTML of a response (`cy.request`, never the hydrated DOM — several findings in this work are precisely about content that exists after hydration and not before, and a hydrated-DOM assertion would pass on exactly the bug it was written to catch). Covering a new route is a row, not a new test; the shape follows the existing search spec.

The link-preview-image reachability assertion is owned by #511 — the same assertion must not exist twice. This ticket asserts titles and descriptions only.

## Acceptance criteria

- [ ] Course page title names the course; professor page title names the professor; both differ from the root default
- [ ] Descriptions carry that page's own rating and review figures, distinct per URL
- [ ] Root title template in place; per-route titles append the brand without repeating it at call sites
- [ ] Missing course/professor yields a not-found title, not a throw
- [ ] One-slot ownership comment present at each slot owning metadata
- [ ] Metadata fetch shares the request-scoped cache with the render — one query per request
- [ ] Table-driven e2e rows asserting titles and descriptions against raw HTML, resolving URLs from the head rather than fixed paths (build-generated hashes change every build)
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- #524
- #525

### LINKED ISSUE #527: [SEO] P0: Publish a sitemap and a robots file
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Sitemap and robots)

## What to build

A course nobody currently links to may never be indexed: there is no sitemap and no robots file, so discovery depends entirely on in-page links — while the crawler is free to spend its budget on the API endpoint, the authentication screens and the submission form.

Both files are framework conventions at the app root; no new dependency, no route handler.

The sitemap enumerates every course and professor from the two existing public catalogue procedures, plus the static routes and public roadmaps. The roadmap listing is cursor-paginated, so the sitemap pages through it rather than assuming one call returns the set. It is cached rather than recomputed per request — a sitemap rebuilt on every crawler visit is a database load problem disguised as an SEO improvement. A single sitemap is correct today; chunking is only required past 50,000 URLs, and that threshold is noted rather than pre-built for.

Robots disallows the API endpoint, the authentication screens, the submission form and the site's own search results, and points at the sitemap. On non-production deployments it disallows everything — which is what keeps a preview host from competing with production in the index, and is why this file is generated rather than static. The deployment-environment gate is added to the validated environment schema rather than read raw: the schema already derives from that variable without exposing it, and reaching for it directly would reintroduce the raw-environment-access pattern the metadata-base ticket removes.

## Acceptance criteria

- [ ] `/sitemap.xml` responds and contains a known course URL and a known professor URL
- [ ] Sitemap sources are the existing public procedures; the roadmap listing paged through its cursor
- [ ] Sitemap cached, not recomputed per request
- [ ] `/robots.txt` responds, disallows the API, auth, submit and search paths, and references the sitemap
- [ ] Non-production robots disallows everything
- [ ] Deployment environment read through the validated schema
- [ ] e2e rows for both files added to the shared seam
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- #525

### LINKED ISSUE #528: [SEO] P0: Canonicalise the parameterised routes
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Canonicals)

## What to build

Filtering a course page by professor, sorting it, or paging through search produces a distinct URL for every combination, and none of them declares which URL is the real one — so the ranking signals a page earns are split across its own variants. Every route that accepts query parameters declares a canonical.

The rule: **a parameter that narrows, reorders or pages a view of an existing page is dropped from the canonical; a parameter without which the URL names nothing is kept.** The test is not "does the rendered content differ" — a filtered view's content does differ, and that is exactly the case being consolidated — it is "is this a subset or ordering of a page that already has its own URL".

The case most likely to be got backwards, stated explicitly: the course page's professor parameter genuinely renders different reviews — and still canonicalises to the bare course URL, because it is a subset of that page rather than a page of its own. The same holds for sort and filter. The search route's query parameter is the opposite case and would be kept, but that route is excluded from indexing anyway, so its canonical is a formality.

## Acceptance criteria

- [ ] Canonical present on every route that accepts query parameters
- [ ] Course page canonical drops the professor, filter and sort parameters — asserted with them present in the requested URL
- [ ] Identifying parameters kept where they name distinct content
- [ ] e2e rows asserting canonical presence and parameter stripping against raw HTML
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- #525
- #526

### LINKED ISSUE #529: [SEO] P1: Structured data for courses, professors, breadcrumbs and the site
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Structured data; Breadcrumbs)

## What to build

A course page becomes eligible for the star ratings that make it stand out in a result list. Structured data is emitted as `application/ld+json` from server components, generated from the same procedures that feed the preview image so the numbers in a search result cannot diverge from the numbers on the page:

- Course pages: the course, with an aggregate rating
- Professor pages: the person, with an aggregate rating
- Both: a breadcrumb trail
- Home: the site, with a search action

The rating scale is 1–5, taken from the rating input's declared maximum, and is declared explicitly in the aggregate rating rather than left to be inferred.

**The one branch that needs a test rather than a type-check: a course or professor with zero reviews omits the aggregate rating entirely.** A zero rating is below the declared minimum, invalid against the vocabulary, and grounds for the whole rich result being dropped — an unrated course must not be presented as a badly rated one. No seed change is needed: eleven of the fifteen seeded courses carry no reviews (`MGMT214` is one).

Breadcrumbs are generated server-side from the records the metadata already fetches — deliberately not from the visual breadcrumb component, which on precisely the routes worth marking up gates its names on a client-side query having succeeded and degrades to a bare "Home" link in the server HTML. The visual component is left alone; #520 tracks it separately under #502.

## Acceptance criteria

- [ ] Course and professor pages emit parseable JSON-LD of the expected types, with the rating scale declared
- [ ] A zero-review course omits the aggregate rating — asserted against the existing seed, no seed change
- [ ] Breadcrumb structured data present in the raw HTML of course and professor routes
- [ ] Home page emits site and search-action structured data
- [ ] Numbers sourced from the same procedures that feed the preview image
- [ ] e2e rows added to the shared seam
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- #526

### LINKED ISSUE #530: [SEO] P1: Mark private and utility routes non-indexable
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Private and authentication routes)

## What to build

A student's private timetable and personal roadmap are indexable today, and so are the authentication screens, the submission form and the site's own search results. Each group is marked at the highest shared point available, so a route added later inherits the setting rather than needing its own:

- The six authentication screens are covered by the existing auth layout — one export, six routes
- The submission form, the site search route and the personal roadmap route are server components and export the directive directly
- The timetable route is a client component at its segment root and so cannot export metadata. It gets a sibling server layout carrying only the directive and its children — deliberately not a restructure of the 686-line client page, which a metadata change has no reason to become

The two share routes already carry the directive and are not modified.

## Acceptance criteria

- [ ] Timetable, personal roadmap, submission form, search and all six authentication screens carry the non-indexing directive in raw HTML
- [ ] Markings sit at the highest shared point per group
- [ ] Content routes do not carry the directive — asserted, not assumed
- [ ] The timetable page component itself untouched
- [ ] e2e rows added to the shared seam
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- None (can start immediately)

### LINKED ISSUE #531: [SEO] P1: Complete Open Graph and add Twitter cards
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Remaining items, Open Graph and Twitter)

## What to build

A shared link renders a proper card everywhere: the root open-graph block gains its image and its URL, and a Twitter card block is added with the large-image type, so a share looks the same on platforms that read either set of tags. Per-route open-graph titles and descriptions flow from the per-route metadata.

The locale is checked against the accepted open-graph locale list before it is changed: that list is a fixed enumeration, not any well-formed language tag, and an unrecognised locale is worse than a merely imprecise one. If the Singapore locale is not on the list, the closest accepted value is used; the current value is the safe status quo. The decision is recorded either way.

The link-preview image route currently returns a server error on every request — #511 owns that fix and the assertion that catches it. This ticket is about the tags and does not duplicate that assertion.

## Acceptance criteria

- [ ] Root open-graph block declares its image and its URL
- [ ] Twitter card block present with the large-image type
- [ ] Locale confirmed against the accepted list before changing, or left unchanged — decision recorded
- [ ] Per-route open-graph inherits titles and descriptions from the per-route metadata
- [ ] e2e rows added to the shared seam
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- #525
- #526

### LINKED ISSUE #532: [SEO] P2: Name course links and fix image alt text
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Remaining items, Anchor text and Alt text; Further Notes — the missed finding)

## What to build

A list of course links is legible before it is followed, and to a screen reader it is not read out as "course, course, course". The course link's anchor text carries the descriptive name currently hidden in a tooltip, not just the code.

The accessible-name override is removed rather than worked around: the link component sets an aria-label that overrides the accessible name with the literal string "course" for every course link on the page. A longer link text is invisible to assistive technology while an overriding label is present, so the two are fixed together — the anchor-text fix alone would be defeated by the attribute.

Image alt text: the announcement card's illustration is either given a genuinely descriptive alt or an empty one marking it decorative — the decision is between those two, not a longer generic string. The link-preview alt text becomes route-specific alongside the per-route metadata.

## Acceptance criteria

- [ ] Course link accessible name includes the course name, verified with the override gone
- [ ] No aria-label override remains on course links
- [ ] Announcement image alt is descriptive or empty-decorative — decided and applied
- [ ] Link-preview alt is route-specific
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- None (can start immediately)

### LINKED ISSUE #533: [SEO] P2: Complete the icon set and add a web manifest
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Remaining items, Icons and manifest)

## What to build

The site is installable and renders correctly when saved to a home screen: the icon set is completed through the framework's icon file conventions alongside the existing favicon, and a web manifest is added the same way. No new dependency.

Blocked by #506, which replaces the favicon through the same icon conventions — two tickets on the same files would collide. This one builds on what lands there.

## Acceptance criteria

- [ ] Icon conventions complete — the set the platform expects, not only the favicon
- [ ] Web manifest present as a file convention
- [ ] Icons and manifest serve correct content types
- [ ] No conflict with the favicon #506 lands
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- #506

### LINKED ISSUE #534: [SEO] P2: Settle redirect permanence and the 404 title
## Parent

Parent issue: #503 · Source spec: `.scratch/specs/seo.md` (Implementation Decisions — Remaining items, Redirects; the not-found title)

## What to build

An old `/reviews` bookmark lands permanently on the page that replaced it, so the destination keeps the ranking the old URL earned — the redirect has been stable, and a temporary redirect withholds the signal it should be passing. The verify-page redirect deliberately stays temporary: it is conditional on a missing query parameter, not a settled move.

A page that does not exist says so in its title, so a mistyped URL is obvious from the browser tab.

## Acceptance criteria

- [ ] `/reviews` redirect made permanent; the verify-page redirect still temporary — both asserted, not just unchanged by accident
- [ ] The 404 page carries a title
- [ ] Standard gates pass: typecheck, lint, format, unit tests, build

## Blocked by

- None (can start immediately)

### PR DESCRIPTION
**Pull request by zekkv**
closes #503
closes #524
closes #525
closes #526
closes #527
closes #528
closes #529
closes #530
closes #531
closes #532
closes #533
closes #534

## Context

The site shipped with default Next.js metadata: root-only titles, no canonicals, no sitemap or robots, and no structured data. Course, professor and roadmap pages could not compete for the searches they should own.

## Changes

- Per-route titles, descriptions and canonicals for course, professor and roadmap pages
- `robots.txt` (`src/app/robots.ts`) with preview hosts blanket-disallowed and private/auth/utility routes disallowed
- `sitemap.xml` (`src/app/sitemap.ts`), statically generated at build from a session-less tRPC caller
- JSON-LD: Course, Person, Breadcrumb and WebSite blocks with a declared rating scale; zero-review entities omit the rating entirely
- Private, auth and utility routes noindex; `/reviews` redirect made permanent (308)
- Removes the dead AnnouncementCarousel and `public/cat.webp`; drops aria-label overrides that hid real link names

## Implementation Details

- `@reviews` is the single metadata owner for course/professor routes; the constraint is documented as comments at each slot, because parallel slots merge in traversal order and the last writer wins
- JSON-LD is escaped against `</script>` breakout (`src/common/components/json-ld.tsx`) and covered by unit tests
- Course JSON-LD carries the Google-required `description`, reused from `courseDescription()`

## How to Test

- Unit: `bun run test` (seo, pageDescriptions, json-ld suites)
- E2E: `cypress/e2e/2-reviews/metadata.cy.js` asserts raw-HTML titles, canonicals, noindex, OG/Twitter tags and parsed JSON-LD per type
- Manual: check `/robots.txt` and `/sitemap.xml` on a preview deployment

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) document
- [x] I have tested the changes
- [x] _(where applicable)_ I have added tests to cover my changes
- [x] _(where applicable)_ I have updated the documentation accordingly

## Summary by CodeRabbit

- **New Features**
  - Added richer SEO metadata, canonical URLs, Open Graph/Twitter details, structured data, robots rules, and sitemap generation.
  - Added descriptive course and professor page metadata, including ratings and review summaries.
  - Added environment configuration for site URLs and deployment-specific crawling behavior.
- **Bug Fixes**
  - Corrected the reviews redirect to use a permanent redirect.
  - Improved Open Graph image generation and accessible labeling for profile and review elements.
- **Documentation**
  - Expanded environment-variable guidance and corrected a contribution-guide link.
- **Tests**
  - Added comprehensive SEO, metadata, structured-data, and description coverage.
- **Removals**
  - Removed the home-page announcement carousel.

### CODE REVIEW

File: src/common/tools/trpc/react.tsx [hunk 1]
```diff
@@ -74,6 +75,7 @@ export function TRPCReactProvider(props: { children: React.ReactNode }) {
 
 function getBaseUrl() {
   if (typeof window !== "undefined") return window.location.origin;
-  if (process.env.VERCEL_URL) return `https://${process.env.VERCEL_URL}`;
-  return `http://localhost:${process.env.PORT ?? 3000}`;
+  // A client-safe validated value: `@/env` proxies server keys in the browser
+  // bundle, and this module is `"use client"`.
+  return env.NEXT_PUBLIC_SITE_URL;
```
[the remark this answers is not in the document]
Comment (zekkv) [reply src/common/tools/trpc/react.tsx:80]: False positive. Vercel auto-provides the framework-prefixed system env vars for Next.js at both build and runtime for production and preview deployments (https://vercel.com/docs/environment-variables/framework-environment-variables), so NEXT_PUBLIC_VERCEL_URL, NEXT_PUBLIC_VERCEL_ENV and NEXT_PUBLIC_VERCEL_PROJECT_PRODUCTION_URL are all set and the fallback chain in env.ts resolves to the deployment URL. Locally, http://localhost:3000 is the intended target. Nothing SSR-fetches through this client either (no suspense/prefetch), so the server branch is dormant.

File: src/common/tools/seo.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  roadmaps: { id: string }[];
+};
+
+export function buildSitemap(
+  baseUrl: string,
+  { courses, professors, roadmaps }: SitemapSources,
+): MetadataRoute.Sitemap {
+  // A single sitemap holds at most 50,000 URLs; split into a sitemap index
+  // before the catalogue reaches that, not after.
+  return [
```
[the remark this answers is not in the document]
Comment (zekkv) [reply src/common/tools/seo.ts:51]: Valid premise (50k limit, unbounded queries), but deferred deliberately: the catalogue is ~38 URLs today, and the ceiling plus upgrade path are documented at seo.ts:49-50 — split into a sitemap index before the catalogue reaches that, not after. Adding generateSitemaps now would change the /sitemap.xml surface for no present benefit. Recorded as future work; if the catalogue approaches the limit this should land first.
